presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveStorageFormats.java

```diff
@@ -430,6 +434,83 @@ private void runTimestampQueries(String tableName, List<TimestampAndPrecision> d
+                tableName,
+                format.getStoragePropertiesAsSql()));
+
+        // Insert in a loop because inserting with UNION ALL sometimes makes values invisible to Presto
+        for (TimestampAndPrecision entry : TIMESTAMPS_FROM_HIVE) {
+            onHive().executeQuery(format(
+                    "INSERT INTO %1$s\n"
+                            // insert with SELECT because hive bans array/map/struct functions in VALUES
+                            + "SELECT\n"
+                            + "%3$s\n"
+                            + ",array(%2$s)\n"
+                            + ",map(%2$s, %2$s)\n"
+                            + ",named_struct('col', %2$s)\n"
+                            + ",array(map(%2$s, named_struct('col', array(%2$s))))\n"
+                            // some hive versions don't allow INSERT from SELECT without FROM
+                            + "FROM dummy\n",
+                    tableName,
+                    format("TIMESTAMP '%s'", entry.getWriteValue()),
+                    entry.getId()));
+        }
+
+        for (HiveTimestampPrecision precision : HiveTimestampPrecision.values()) {
+            setSessionProperty(onPresto().getConnection(), "hive.timestamp_precision", precision.name());
+            assertThat(onPresto()
+                    .executeQuery(format(
+                            "SELECT\n"
+                                    + "id\n"
+                                    + ",arr[1]\n"
+                                    + ",map_entries(map)[1][1]\n" // key
+                                    + ",map_entries(map)[1][2]\n" // value
+                                    + ",row.col\n"
+                                    + ",map_entries(nested[1])[1][1]\n" // key
+                                    + ",map_entries(nested[1])[1][2].col[1]\n" // value
+                                    + "FROM %s\n"
+                                    + "ORDER BY id\n",
+                            tableName)))
+                    .as("timestamp containers on %s", format.getName().toLowerCase(Locale.ENGLISH))
+                    .containsExactly(TIMESTAMPS_FROM_HIVE.stream()
+                            .sorted(comparingInt(TimestampAndPrecision::getId))
+                            .map(e -> new Row(Lists.asList(
+                                    e.getId(),
+                                    nCopies(6, Timestamp.valueOf(e.getReadValue(precision))).toArray())))
+                            .collect(toList()));
+
+            for (TimestampAndPrecision entry : TIMESTAMPS_FROM_HIVE) {
+                if (entry.isRounded(precision)) {
+                    // Check rounding in the array column
+                    assertThat(onPresto()
+                            .executeQuery(format(
+                                    "SELECT id FROM %2$s WHERE id = %3$s AND TIMESTAMP '%4$s' %1$c arr[1]",
+                                    entry.isRoundedUp(precision) ? '<' : '>',
```

Comment:
That's hard to read. Why not just read the value and check it is exactly as expected?

See also https://github.com/prestosql/presto/pull/6297

Reply:
I originally had based this on `runTimestampQueries`. Now that the rounding checks have been removed there, I've removed this as well.